# Peptide Coregulation Analysis

This repository contains the code and analysis scripts used in the paper:

Zhongzhi Sun, Zhibin Ning, Qing Wu et al., Peptide abundance correlations in metaproteomics enhance taxonomic and functional analysis of the human gut microbiome
(DOI: [to be added]).

## Repository Structure

- `Manuscript_Data_Analysis.Rmd`: Main data analysis scripts
- `Machine_Learning_Peptide_Tax_Assign.Rmd`: Machine learning approach for peptide taxonomic assignment
- `Manuscript_Figures.Rmd`: Code for generating manuscript figures
- `Manuscript_Supplementary_Figures.Rmd`: Code for generating supplementary figures

## Requirements

To run these analyses, you will need:
- R version 4.0 or higher
- RStudio for running R Markdown files

## Installation

1. Clone this repository
2. Install required R packages (see `requirements.R`)

## Usage

Each R Markdown file can be run independently in RStudio. The files should be run in the following order:
1. Manuscript_Data_Analysis.Rmd
2. Machine_Learning_Peptide_Tax_Assign.Rmd
3. Manuscript_Figures.Rmd
4. Manuscript_Supplementary_Figures.Rmd

## Citation

If you use this code in your research, please cite:

[Citation information to be added after publication]

## License

This project is licensed under the MIT License - see the LICENSE file for details.

## Contact
Zhongzhi Sun
zsun036@uottawa.ca